interfaces/prompting/patterns/variant.go

```diff
@@ -0,0 +1,428 @@
+				components[len(components)-1] = component{compType: compSeparatorDoublestar}
+			} else {
+				// Reduce to * since previous component is not a separator
+				addGlobstar()
+			}
+		case '*':
+			reducePrevDoublestar()
+			consumeText()
+			addGlobstar()
+		case '\\':
+			r2, _, err := rr.ReadRune()
+			if err != nil {
+				// Should be impossible, we just rendered this variant
+				return PatternVariant{}, errors.New(`internal error: trailing unescaped '\' character`)
+			}
+			switch r2 {
+			case '*', '?', '[', ']', '{', '}', '\\':
+				runes = append(runes, r, r2)
+			default:
+				// do not add r to runes if it's unnecessary
+				runes = append(runes, r2)
+			}
+		case '[', ']', '{', '}':
+			// Should be impossible, we just rendered this variant
+			return PatternVariant{}, fmt.Errorf(`internal error: unexpected unescaped '%v' character`, r)
+		default:
+			runes = append(runes, r)
+		}
+	}
+
+	consumeText()
+
+	if prevComponentsAre([]componentType{compSeparatorDoublestar, compSeparator, compGlobstar}) {
+		// If components end with /**/*, strip trailing /*
+		components = components[:len(components)-2]
+	}
+
+	// Add terminal marker or convert existing doublestar to terminal doublestar
+	if prevComponentsAre([]componentType{compSeparatorDoublestar, compSeparator}) {
+		components = components[:len(components)-2]
+		components = append(components, component{compType: compSeparatorDoublestarSeparatorTerminal})
+	} else if prevComponentsAre([]componentType{compSeparatorDoublestar}) {
+		components[len(components)-1] = component{compType: compSeparatorDoublestarTerminal}
+	} else {
+		components = append(components, component{compType: compTerminal})
+	}
+
+	var variantBuf strings.Builder
+	var regexBuf strings.Builder
+	regexBuf.WriteRune('^')
+	for _, c := range components {
+		variantBuf.WriteString(c.String())
+		regexBuf.WriteString(c.componentRegex())
+	}
+	regexBuf.WriteRune('$')
+	regex := regexpMustCompileLongest(regexBuf.String())
+
+	v := PatternVariant{
+		variant:    variantBuf.String(),
+		components: components,
+		regex:      regex,
+	}
+
+	return v, nil
+}
+
+// regexpMustCompileLongest compiles the given string into a Regexp and then
+// calls Longest() on it before returning it.
+func regexpMustCompileLongest(str string) *regexp.Regexp {
+	re := regexp.MustCompile(str)
+	re.Longest()
+	return re
+}
+
+// Need to escape any unescaped literal "⁑" runes before we use that symbol to
+// indicate the presence of a "**" doublestar.
+var doublestarEscaper = regexpMustCompileLongest(`((\\)*)⁑`)
+
+// Need to replace unescaped "**", but must be careful about an escaped '\\'
+// before the first '*', since that doesn't escape the first '*'.
+var doublestarReplacer = regexpMustCompileLongest(`((\\)*)\*\*`)
+
+// prepareVariantForParsing escapes any unescaped '⁑' characters and then
+// replaces any unescaped "**" with a single '⁑' so that doublestars can be
+// identified without needing to look ahead whenever a '*' is seen.
```

Comment:
it's questionable whether this is clearer than one char lookahead

Reply:
Fair enough. I wanted the switch cases when reading runes in `ParsePatternVariant` to be unique, so that each individual rune could be treated as a "token" (with the exception of non-special runes, which are grouped via `consumeText`). In previous iterations of this PR, doing lookahead got very messy, without clear distinctions between tokenization and parsing

Since it is just one character, I'm open to switching to lookahead, but for now I'll leave it as-is.

Reply:
this can be a follow up but I think it would be a good idea to avoid needing this